SimMuon/MCTruth/src/MuonAssociatorByHitsHelper.cc

```diff
@@ -0,0 +1,1154 @@
+#include "DataFormats/DetId/interface/DetId.h"
+#include "DataFormats/MuonDetId/interface/MuonSubdetId.h"
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "DataFormats/SiStripDetId/interface/StripSubdetector.h"
+#include "DataFormats/SiPixelDetId/interface/PixelSubdetector.h"
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
+#include "DataFormats/DTRecHit/interface/DTRecSegment4D.h"
+#include "DataFormats/CSCRecHit/interface/CSCSegment.h"
+#include <sstream>
+
+using namespace reco;
+using namespace std;
+
+MuonAssociatorByHitsHelper::MuonAssociatorByHitsHelper (const edm::ParameterSet& conf) :  
+  includeZeroHitMuons(conf.getParameter<bool>("includeZeroHitMuons")),
+  acceptOneStubMatchings(conf.getParameter<bool>("acceptOneStubMatchings")),
+  UseTracker(conf.getParameter<bool>("UseTracker")),
+  UseMuon(conf.getParameter<bool>("UseMuon")),
+  AbsoluteNumberOfHits_track(conf.getParameter<bool>("AbsoluteNumberOfHits_track")),
+  NHitCut_track(conf.getParameter<unsigned int>("NHitCut_track")),
+  EfficiencyCut_track(conf.getParameter<double>("EfficiencyCut_track")),
+  PurityCut_track(conf.getParameter<double>("PurityCut_track")),
+  AbsoluteNumberOfHits_muon(conf.getParameter<bool>("AbsoluteNumberOfHits_muon")),
+  NHitCut_muon(conf.getParameter<unsigned int>("NHitCut_muon")),
+  EfficiencyCut_muon(conf.getParameter<double>("EfficiencyCut_muon")),
+  PurityCut_muon(conf.getParameter<double>("PurityCut_muon")),
+  UsePixels(conf.getParameter<bool>("UsePixels")),
+  UseGrouped(conf.getParameter<bool>("UseGrouped")),
+  UseSplitting(conf.getParameter<bool>("UseSplitting")),
+  ThreeHitTracksAreSpecial(conf.getParameter<bool>("ThreeHitTracksAreSpecial")),
+  dumpDT(conf.getParameter<bool>("dumpDT"))
+{
+  edm::LogVerbatim("MuonAssociatorByHitsHelper") << "constructing  MuonAssociatorByHitsHelper" << conf.dump();
+
+  // up to the user in the other cases - print a message
+  if (UseTracker) edm::LogVerbatim("MuonAssociatorByHitsHelper")<<"\n UseTracker = TRUE  : Tracker SimHits and RecHits WILL be counted";
+  else edm::LogVerbatim("MuonAssociatorByHitsHelper") <<"\n UseTracker = FALSE : Tracker SimHits and RecHits WILL NOT be counted";
+  
+  // up to the user in the other cases - print a message
+  if (UseMuon) edm::LogVerbatim("MuonAssociatorByHitsHelper")<<" UseMuon = TRUE  : Muon SimHits and RecHits WILL be counted";
+  else edm::LogVerbatim("MuonAssociatorByHitsHelper") <<" UseMuon = FALSE : Muon SimHits and RecHits WILL NOT be counted"<<endl;
+  
+  // check consistency of the configuration when allowing zero-hit muon matching (counting invalid hits)
+  if (includeZeroHitMuons) {
+    edm::LogVerbatim("MuonAssociatorByHitsHelper") 
+      <<"\n includeZeroHitMuons = TRUE"
+      <<"\n ==> (re)set NHitCut_muon = 0, PurityCut_muon = 0, EfficiencyCut_muon = 0"<<endl;
+    NHitCut_muon = 0;
+    PurityCut_muon = 0.;
+    EfficiencyCut_muon = 0.;
+  }
+
+}
```

Comment:
I was wrong on the above. It is not possible to add ```crossingframe``` since that would require getting more data from the Event. As ```crossingframe``` is only used if ```dumpInputCollections``` is true, would ```dumpInputCollections``` still be useful without ```crossingframe```?